### DISCUSSION TITLE
`exec`ing from Singularity-Container

### DISCUSSION THREAD

**Discussion by cstenkamp** [frankier/skelshop#35] [Q&A]
Hello Frankie,
in https://github.com/frankier/skelshop/blob/master/contrib/slurm/conv_2017.sh#L13-L16, you give an example on how to run 
arbitrary commands from inside the singularity container, however I cannot reproduce that. Running for example 

```
singularity exec ./skelshop_latest.sif python /opt/skelshop/skelshop dump ~/small_dataset/Bethenny_Frankel_01.mp4 ~/tmp.h5 --mode BODY_25_ALL 
```
simply tells me "Illegal Instruction", which is unfortunately not very verbose. I also tried various variants of that, but never got it to work. What would I need to do to run arbitrary skelshop-commands from inside the container?

(and also, how does that work? When looking at `singularity exec ./skelshop_latest.sif pip list`, skelshop is listed as package. How did it become that? And why can I simply run `python` from the container, and not `poetry run python`? And why doesn't something like `python -m /opt/skelshop --help` work?) 
Thanks for your help, and again, sorry for all the stupid questions!

**Comment by frankier** [comment]
Supposing you're running on x86-64, illegal instruction might be to do with some processor specific extensions e.g. AVX512 being run. That said the Singularity container is currently CUDA only so you always need to run with `--nv`. That would case a different error however. At any rate you can use the `docker://frankierr/skelshop:focal_cpu` Docker container for CPU usage.

If we can find out what instruction extensions the different images/libraries require, we can at least add checks beforehand so people get a more useful error message.

Unfortunately there's still some out of date stuff in this repository -- so it's not stupid at all -- unfortunately I keep having to change things around to get everything to work together.

I'm considering getting rid of the Singularity container altogether and just advising using the Docker container through Singularity. I think this would make things a bit simpler.

That's odd it's not listed as a package for you because it is for me. Using python -m skelshop is now the recommended way because it work across different environments. The reason poetry is not needed is because the packages are all installed globally. This is done with this line:

https://github.com/frankier/skelshop/blob/114976f2b462ec55ee1b46d8faccf53a7dbfa3b3/Dockerfile#L41

**Comment by cstenkamp** [comment]
Okay, I noticed by now that my errror was that I was only on the hpclogin-machine. If i ssh to eg. hpc4 I don't get the Illegal instructions anymore, and it also works with the singularity container.
Regarding your last paragraph: I said it was listed, I was just surprised why, so thanks for telling me :)
Also, do you know if there are any machines with GPU I can ssh to? hpc4 works, but not for eg. dumping, as I need a gpu-machine there...

**Reply by frankier** [reply]
Hi. Sorry I think I missed this before. You can use SLURM to get a GPU machine usually. Is this Case we're talking about? If so you can either use the gpu partition and gres and use sbatch maunally or use the Snakemake/singslurm approach and put the same stuff in your cluster configuration JSON.

**Reply by cstenkamp** [reply]
Hm, but you don't know of any machine I can simply ssh to? Sounds easier for testing

**Reply by frankier** [reply]
https://sites.google.com/a/case.edu/hpcc/jobs/interactive-session
